sklearn/tree/_tree.pyx

```diff
@@ -1132,3 +1132,296 @@ cdef class Tree:
         Py_INCREF(self)
         arr.base = <PyObject*> self
         return arr
+
+# =============================================================================
+# Build Pruned Tree
+# =============================================================================
+
+cpdef build_pruned_tree_ccp(
+    Tree tree, # OUT
+    Tree orig_tree,
+    DOUBLE_t ccp_alpha):
+    """Builds a pruned tree with Minimal Cost-Complexity Pruning.
+
+    Builds a pruned tree from the original tree. The values and nodes from the
+    original tree are copied into the pruned tree.
+
+    Parameters
+    ----------
+    tree : Tree
+        Location to place the pruned tree
+    orig_tree : Tree
+        Original tree
+    ccp_alpha : positive double
+        Complexity parameter. The subtree with the largest cost complexity
+        that is smaller than ``ccp_alpha`` will be choosen. By default,
+        no pruning is performed.no pruning is performed.
+    """
+
+    cdef:
+        SIZE_t i
+        SIZE_t n_nodes = orig_tree.node_count
+        # prior probability using weighted samples
+        DOUBLE_t[:] weighted_n_node_samples = orig_tree.weighted_n_node_samples
+        DOUBLE_t total_sum_weights = weighted_n_node_samples[0]
+        DOUBLE_t[:] impurity = orig_tree.impurity
+        # weighted impurity of each node
+        DOUBLE_t[:] r_node = np.empty(shape=n_nodes, dtype=np.float64)
+
+        SIZE_t[:] child_l = orig_tree.children_left
+        SIZE_t[:] child_r = orig_tree.children_right
+        SIZE_t[:] parent = np.zeros(shape=n_nodes, dtype=np.intp)
+        unsigned char[:] leaves_in_subtree = np.zeros(
+            shape=n_nodes, dtype=np.uint8)
+
+        # Only uses the start and parent variables
+        Stack stack = Stack(INITIAL_STACK_SIZE)
+        StackRecord stack_record
+        int rc = 0
+        SIZE_t node_idx
+
+        SIZE_t[:] n_leaves = np.zeros(shape=n_nodes, dtype=np.intp)
+        DOUBLE_t[:] r_branch = np.zeros(shape=n_nodes, dtype=np.float64)
+        DOUBLE_t current_r
+        SIZE_t leaf_idx
+        SIZE_t parent_idx
+
+        # candidate nodes that can be pruned
+        bint[:] candidate_nodes = view.array(
+            shape=(n_nodes, ), itemsize=sizeof(bint), format="i")
+        # nodes in subtree
+        bint[:] in_subtree = view.array(
```

Comment:
out of curiosity, what makes you choose between `np.array`, `cython.view.array` and `cpython.array` (if you were to use it)? Here you're mixing different memory allocation methods which makes me think it's intentional.

Reply:
My original intention was to use a `bint` because it was used throughout the codebase for a bool. From the Cython documentation, it looks like `bint` becomes normal ints. I intend to change all the `view.array` calls to `np.zeros(shape=n_nodes, dtype=np.uint8)` I missed this one.